### PR TITLE
Added Otp Sender #18

### PR DESCRIPTION
**Pull request by artfred16**
Added Otp Sender#18
Added Otp Verification #7
Resend Verification #6
Refactor user registration #4

**To Test**
1. compile vue `npm run dev`
2. serve php `artisan serve --port 22222`
3. visit at [localhost:2222](localhost:2222)
4. click signup
5. enter mobile number then click `verify`
image

6. type the code above page `*NOTE: tempory to display code`
image

7. register by entering details
image

8. clicking register should display user account
image

### CODE REVIEW

**Comment by jerud3v** [comment]
Please correct me if I'm wrong, but shouldn't be the `verified` column be updated upon verification?

**Comment by artfred16** [comment]
> Please correct me if I'm wrong, but shouldn't be the `verified` column be updated upon verification?
> 
> 

@jerud3v  We changed are workflow to only register after mobile has been verified. thus verified field will not be used to verify mobile but insteal will be used to verify email

File: app/Http/Controllers/Api/MobileValidationController.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+use App\Http\Controllers\Controller;
+use App\User;
+use Illuminate\Http\Request;
+use Illuminate\Http\Response;
+use Illuminate\Support\Facades\Cache;
+use Illuminate\Support\Facades\Crypt;
+use Libraries\Otp;
+
+class MobileValidationController extends Controller {
+	private const IDENTITY_VALIDITY = 900;
```
Comment (kristher1619) [inline app/Http/Controllers/Api/MobileValidationController.php:13]: Please change to CODE_EXPIRATION = 900 // Add the human readable time as comment

File: app/Http/Controllers/Api/VerificationController.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+use App\Http\Controllers\Controller;
+use Illuminate\Support\Facades\Hash;
+use Libraries\Otp;
+use Illuminate\Http\Response;
+use Illuminate\Support\Facades\Cache;
+use Illuminate\Support\Facades\Crypt;
+
+class VerificationController extends Controller {
+
+	private const EXP = 1800;
```
Comment (kristher1619) [inline app/Http/Controllers/Api/VerificationController.php:14]: change to EXPIRATION // with human readable time equivalent as comment

File: resources/js/components/RegisterForm.vue [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        created() {
+            this.form.confirmation_token = this.token;
+        },
+        methods: {
+            async register() {
+                this.errorMessage = ''
+                await this.$axios.get('/sanctum/csrf-cookie')
+
+                this.form.post('/api/register').then((response) => {
+                    window.location.replace('/account-update')
```
Comment (kristher1619) [inline resources/js/components/RegisterForm.vue:83]: change to update-profile

File: tests/Feature/MobileValidationTest.php [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+		$result = $this->post('api/validate', $request)
+		     ->assertRedirect();
+		$location = $result->headers->get('location');
+		$this->assertStringContainsString('/verify', $location);
+	}
+
+	/**
+	 * @test
+	 */
+	public function it_should_return_error_if_mobile_was_previously( )
```
Comment (kristher1619) [inline tests/Feature/MobileValidationTest.php:42]: this is incomplete sentence, was_previously what?

File: app/Http/Controllers/Api/MobileValidationController.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	public function store( Request $request ) {
+		$request->validate( [
+			'mobile_number' => [
+				'required',
+				function ( $attribute, $value, $fail ) use ( $request ) {
+					if ( User::whereMobileNumber( $value )->first() ) {
+						$fail( 'Mobile number was previously registered.'
+						       . 'Please user another number or logon to your account' );
+					}
+				}
```
Comment (kristher1619) [inline app/Http/Controllers/Api/MobileValidationController.php:25]: why not just use the built in validation exists?

File: routes/web.php [hunk 6]
```diff
@@ -26,13 +26,10 @@
...
-    Route::get('verify', function () {
-        // Temporary page
-        return 'Verify';
-    });
+    Route::get('signup', function () { return view('signup'); });
+	Route::get('verify', function () { return view('verify'); });
+	Route::get('register', function () { return view('create-account'); });
+	Route::get('account-update', function () {
+		return App\User::latest()->first();
+	});
```
Comment (kristher1619) [inline routes/web.php:34]: Use Controller for these codes

File: resources/views/verify.blade.php [hunk 7]
```diff
@@ -0,0 +1,12 @@
+@extends('layouts.master', ['bodyClass' => 'auth'])
+
+@section('content')
+    @php
+        use \Illuminate\Support\Facades\Cache;
+        $key = Illuminate\Support\Facades\Crypt::decrypt(request('token'));
+        $identifier = Cache::get($key);
+        $code = Cache::get('otp-'.$identifier);
+    @endphp
```
Comment (kristher1619) [inline resources/views/verify.blade.php:9]: extract to controller

File: resources/js/components/VerificationForm.vue [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        </div>
+        <form
+            @keydown="form.errors.clear($event.target.name)"
+            @submit.prevent="verify()"
+            class="tw-flex tw-flex-col tw-px-12 tw-items-center"
+        >
+            <div class="form-group">
+                <input
+                    v-model="form.code"
+                    type="password"
```
Comment (kristher1619) [inline resources/js/components/VerificationForm.vue:18]: this should be type="text" so the user can see the code

File: resources/views/verify.blade.php (hunk 7 above)
Comment (artfred16) [reply resources/views/verify.blade.php:9]: this is temporary, used only to display the otp on the page.

File: app/Http/Controllers/Api/VerificationController.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+				function ( $attribute, $value, $fail ) use ($otp) {
+					if ( ! $otp->isValid( $value ) ) {
+						$fail( "Verification code was expired or invalid. Please login or resend verification code" );
+					}
+				}
+			]
+		]);
+
+		$token = Hash::make($key);
+		Cache::put($token, explode('-', $key)[1], self::EXPIRATION);
```
Comment (kristher1619) [inline app/Http/Controllers/Api/VerificationController.php:33]: This extracts the mobile number from the previous key, may I suggest to declare a variable here to make it more readable

**Pull request merged**